Adds section on Advanced Settings (#124915)
elastic/kibana: dev_docs/tutorials/advanced_settings.mdx

---
id: kibDevTutorialAdvancedSettings
slug: /kibana-dev-docs/tutorials/advanced-settings
title: How to register a new advanced setting
summary: Learn how to add and use a new advanced setting
date: 2022-02-07
tags: ['kibana','onboarding', 'dev', 'architecture', 'tutorials']
---

_Note: Advanced settings, uiSettings, settings and config are often used to describe the same concept. While we work toward unifying naming, there will be some inconsistencies until unification is complete._

> UI settings are configurable from the Advanced Settings page in Management and control the behavior of Kibana. When settings are changed from their default, the new value is persisted as a key/value pair in the `config` saved object registered by `core`.

Reviewer: ```suggestion
*Advanced Settings* control the behavior of Kibana. To configure the UI settings, open the main menu, then click *Stack Management > Advanced Settings*.  When settings are changed from their default, the new value is persisted as a key/value pair in the `config` saved object registered by `core`.
```

 - <DocLink id="kibDevTutorialAdvancedSettings" section="client-side-usage" text="Through the client-side `uiSettings` service"/>
 - <DocLink id="kibDevTutorialAdvancedSettings" section="server-side-usage" text="Through the server-side `uiSettings` service"/>

`uiSettings` are registered synchronously during `core`'s setup lifecycle phase. This means that once you add a new advanced setting, you cannot change or remove it without <DocLink id="kibDevTutorialAdvancedSettings" section="registering-migrations" text="registering a migration in core"/>.

### Configuration with Advanced Settings UI

The `uiSettings` service is the programmatic interface to Kibana's Advanced Settings UI. Kibana plugins use the service to extend Kibana UI Settings Management with custom settings for a plugin.

Configuration through the Advanced Settings UI is restricted to users authorised to access the Advanced Settings page. Users who don't have permissions to change these values default to using the csettings configuration defined for the space that they are in. The `config` saved object can be shared between spaces.

### Configuration with UI settings overrides

> When a setting is configured as an override in `kibana.yml`, it will override any other value store in the `config` saved object. The override applies to Kibana as a whole for all spaces and users and the option will be disabled in the Advanced Settings page. We refer to these as "global" overrides.

Reviewer: ```suggestion
When a setting is configured as an override in `kibana.yml`, it overrides any other value store in the `config` saved object. The override applies to Kibana as a whole for all spaces and users, and the option is disabled on the Advanced Settings page. We refer to these as "global" overrides.
```

Note: If an override is misconfigured, it will fail config validation and prevent Kibana from starting up. Validation is, however, limited to value _type_ and not to _key_ (name). For example, if a plugin registers a setting `my_plugin_foo: 42` then declaring an override as

```kibana.yml
uiSettings.overrides:
  my_plugin_foo: "42"
```

> will fail on config validation but the following won't:

Reviewer: ```suggestion
The following override results in a successful config validation:
```

```kibana.yml
uiSettings.overrides:
  my_pluginFoo: 42
```

### Client side usage

On the client, the `uiSettings` service is accessible directly from `core` and the client provides plugins access to the `config` entries stored in Elasticsearch.

<DocCallOut>

>   See [the client-side uiSettings service API docs](https://github.com/elastic/kibana/blob/main/docs/development/core/server/kibana-plugin-core-public.iuisettingsclient.md)

Reviewer: ```suggestion
  Refer to [the client-side uiSettings service API docs](https://github.com/elastic/kibana/blob/main/docs/development/core/server/kibana-plugin-core-public.iuisettingsclient.md)
```

</DocCallOut>

> Here is a basic example for using the `uiSettings` service:

Reviewer: ```suggestion
The following is a basic example for using the `uiSettings` service:
```

    dependencies.expressions.registerFunction(systemPalette);
    this.themeService.init(core.uiSettings);
    this.legacyColorsService.init(core.uiSettings);
    this.palettes = this.paletteService.setup(this.legacyColorsService);

    this.activeCursor.setup();

    return {
      legacyColors: this.legacyColorsService,
      theme: this.themeService,
      palettes: this.palettes,
    };
  }

  public start(): ChartsPluginStart {
    return {
      legacyColors: this.legacyColorsService,
      theme: this.themeService,
      palettes: this.palettes!,
      activeCursor: this.activeCursor,
    };
  }
}

```

### Server side usage

On the server side, `uiSettings` are accessible directly from `core`. The following example shows how to register a new setting with the minimum required schema parameter against which validations are performed on read and write.
The example also shows how plugins can leverage the optional deprecation parameter on registration for handling deprecation notices and renames. The deprecation warnings are rendered in the Advanced Settings UI and should also be added to the Configure Kibana guide.

<DocCallOut>

>   See [the server-side uiSettings service API docs](https://github.com/elastic/kibana/blob/main/docs/development/core/server/kibana-plugin-core-server.iuisettingsclient.md)

Reviewer: ```suggestion
  Refer to [the server-side uiSettings service API docs](https://github.com/elastic/kibana/blob/main/docs/development/core/server/kibana-plugin-core-server.iuisettingsclient.md)
```

    dependencies.expressions.registerFunction(systemPalette);
    core.uiSettings.register({
      [COLOR_MAPPING_SETTING]: {
        name: i18n.translate('charts.advancedSettings.visualization.colorMappingTitle', {
          defaultMessage: 'Color mapping',
        }),
        value: JSON.stringify({
          Count: '#00A69B',
        }),
        type: 'json',
        description: i18n.translate('charts.advancedSettings.visualization.colorMappingText', {
          defaultMessage:
            'Maps values to specific colors in charts using the <strong>Compatibility</strong> palette.',
        }),
        deprecation: {
          message: i18n.translate(
            'charts.advancedSettings.visualization.colorMappingTextDeprecation',
            {
              defaultMessage:
                'This setting is deprecated and will not be supported in a future version.',
            }
          ),
          docLinksKey: 'visualizationSettings',
        },
        category: ['visualization'],
        schema: schema.string(),
      },
    ...
    });
    
    return {};
  }
  
  public start() {
    return {};
  }

  public stop() {}
}  
```

> To keep Kibana fast, `uiSettings` are cached. Any changes that require a cache refresh should use the `requiresPageReload` parameter on registration. 

Reviewer: ```suggestion
For optimal Kibana performance, `uiSettings` are cached. Any changes that require a cache refresh should use the `requiresPageReload` parameter on registration. 
```

**src/plugins/data/server/ui_settings.ts**

```ts
import { i18n } from '@kbn/i18n';
import { schema } from '@kbn/config-schema';
import type { DocLinksServiceSetup, UiSettingsParams } from 'kibana/server';
import { DEFAULT_QUERY_LANGUAGE, UI_SETTINGS } from '../common';

export function getUiSettings(
  docLinks: DocLinksServiceSetup
): Record<string, UiSettingsParams<unknown>> {
  return {
    ...
    [UI_SETTINGS.TIMEPICKER_REFRESH_INTERVAL_DEFAULTS]: {
      name: i18n.translate('data.advancedSettings.timepicker.refreshIntervalDefaultsTitle', {
        defaultMessage: 'Time filter refresh interval',
      }),
      value: `{
      "pause": false,
      "value": 0
      }`,
        type: 'json',
        description: i18n.translate('data.advancedSettings.timepicker.refreshIntervalDefaultsText', {
          defaultMessage: `The timefilter's default refresh interval. The "value" needs to be specified in milliseconds.`,
        }),
        requiresPageReload: true,
        schema: schema.object({
          pause: schema.boolean(),
          value: schema.number(),
        }),
      },
    ...
  }
}
```

### Registering Migrations

> To change or remove a `uiSetting`, the whole `config` Saved Object needs to be migrated. `uiSettings` migrations are declared directly in the service.

Reviewer: ```suggestion
To change or remove a `uiSetting`, you must migrate the whole `config` Saved Object. `uiSettings` migrations are declared directly in the service.
```

> For example, in 7.9.0, `siem` as renamed to `securitySolution` and in 8.0.0, `theme:version` was removed:

Reviewer: ```suggestion
For example, in 7.9.0, `siem` as renamed to `securitySolution`, and in 8.0.0, `theme:version` was removed:
```
